Personal API keys and Zapier integration

Twixes opened the pull request:
## Changes

This adds personal API keys which, combined with `Bearer` auth, allow users to use the API as if logged in normally.

![Personal API keys in Setup](https://user-images.githubusercontent.com/4550621/88557537-9ed22f00-d02a-11ea-87b3-0a5ca522f807.png)
![Personal API key creation](https://user-images.githubusercontent.com/4550621/88554498-e35bcb80-d026-11ea-9a5f-9d7acd858e63.png)

Personal API keys also facilitate read-write integration with external services. This PR covers [Zapier](http://zapier.com/), which works PostHog-side in tandem with a Zapier CLI integration that is uploaded to them: [posthog-zapier](https://github.com/PostHog/posthog-zapier).

![Zapier triggers](https://user-images.githubusercontent.com/4550621/89451446-85349400-d75c-11ea-97d6-4b11c9fce0bc.png)
![Zapier actions](https://user-images.githubusercontent.com/4550621/89451509-9f6e7200-d75c-11ea-87b7-d32e6699e01a.png)

This integration is a paid EE feature and requires that the user is on a license covering the `"zapier"` feature.

## Checklist

- [x] All querysets/queries filter by Team (if applicable)
- [x] Backend tests (if applicable)
- [ ] Cypress E2E tests (if applicable)

Twixes committed:
Add missing migration

Twixes committed:
Add generate_random_token() model util

Twixes committed:
Move PublicTokenAuthentication to utils

Twixes committed:
Make use of generate_random_token

Twixes committed:
Add User.personal_access_token field

Twixes committed:
Add PersonalAccessTokenAuthentication

Twixes committed:
Fix PublicTokenAuthentication

Twixes committed:
Fix migration and auth import

Twixes committed:
Add personal_access_token to user API

Twixes committed:
Update Setup.js

Twixes committed:
Support trailing slash in API

Twixes committed:
Improve PAT auth quality

Twixes committed:
Add django-rest-hooks requirement

Twixes committed:
Update settings.py for rest_hooks

Twixes committed:
Fix django-rest-hooks requirement

Twixes committed:
Bring back API routes with no double trailing slash

Twixes committed:
Rename posthog.api.team to team_user

Twixes committed:
Add API TODO

Twixes committed:
Ad PAT auth with X-PAT HTTP header

Twixes committed:
Replace User.personal_access_token with PersonalAPIKey model

Twixes committed:
Fix PersonalAPIKey max_lengths

Twixes committed:
Describe posthog.models.utils.generate_random_token better

Twixes committed:
Add personal_api_key to API

Twixes committed:
Add authenticate_header to PersonalAPIKeyAuthentication

Twixes committed:
Add hook API endpoint

Twixes committed:
Use django.utils.timezone in place of datetime.datetime

Twixes committed:
Add Personal API Keys to Setup

Twixes committed:
Sort personal_api_keys in ORM

Twixes committed:
Add Action.on_perform()

Twixes committed:
Remove requirements.txt comment

Twixes committed:
Add a

Twixes committed:
Add REST hook tasks

Twixes committed:
Optimize PersonalAPIKeyAuthentication query

Twixes committed:
Add a trailing slash version of /e endpoint

Twixes committed:
Add team field to PersonalAPIKey model

Twixes committed:
Add personal API key support to capture endpoint, get_cached_from_token

Twixes committed:
Reject personal API keys from inactive users

Twixes committed:
Add extra_properties_json field to /capture

Twixes committed:
Improve PAK auth header regex

Twixes committed:
Use custom hook model

Twixes committed:
Deliver hooks

Twixes committed:
Handle action.on_perform

Twixes committed:
Consolidate userLogic in userLogic.tsx

Twixes committed:
Update PersonalAPIKeys.js

Twixes committed:
Make PersonalAPIKey foreign keys read-only

Twixes committed:
Update requirements/dev.txt

Twixes committed:
Make PersonalAPIKeys TSX

Twixes committed:
Fix conflict

Twixes committed:
Fix migration

Twixes committed:
Merge branch 'master' into 906-personal-api-keys-and-zapier

Twixes committed:
Fix minor mishaps

Twixes committed:
Update and fix tests

Twixes committed:
Use CharField of random 32 bits as hook.id

Twixes committed:
Merge branch 'master' into 906-personal-api-keys-and-zapier

Twixes committed:
Fix conflicting migrations

Twixes committed:
Fix ValidationError in HookSerializer.validate_event

timgl commented:
failing tests and probably merge conflicts now as well, but LMK when ready to review

Twixes committed:
Use query param in /api/event/actions ID filtering

Twixes committed:
Rename endpoint `hook` to `hooks`

Twixes committed:
Satisfy mypy

Twixes committed:
Add tests

Twixes committed:
Use DRF serialization in action_defined and annotation_created triggers

Twixes committed:
Merge branch 'master' into 906-personal-api-keys-and-zapier

Twixes committed:
Update migration leafs

Twixes committed:
Make mypy ignore rest_hooks

Twixes committed:
Update Django signal receiver names

Twixes committed:
Update TS dependencies

Twixes commented:
Ahh, [npm down](https://status.npmjs.org/)

Twixes committed:
Revert "Update TS dependencies"

This reverts commit 7fc26fefcdc16e630e1c8fd2c510fd323d97169f.

Twixes commented:
If you want to test Zapier integration locally, create a private integration with https://github.com/PostHog/posthog-zapier. It should be a smooth process.

Twixes committed:
Add field user to Hook model

Twixes committed:
Merge branch 'master' into 906-personal-api-keys-and-zapier

Twixes committed:
Update migration leafs

Twixes committed:
Fix circular import

ELanning commented on frontend/src/scenes/userLogic.tsx:
```diff
@@ -98,5 +118,38 @@ export const userLogic = kea<userLogicType<UserType, EventProperty>>({
+            const user = selectors.user()
+            actions.deletePersonalAPIKeySuccess(user, key)
+        },
+        deletePersonalAPIKeySuccess: ({ key }) => {
```
Because `key.label` is the only property used, consider:
```
deletePersonalAPIKeySuccess(label)
```

Unless more properties of `key` are planning on being used in the future _or_ has better abstractional integrity.

Twixes replied:
I recommend a read through https://kea.js.org/docs/guide/concepts/. It covers always passing payloads in the form of objects.

ELanning replied:
The suggestion wasn't necessarily about wrapping the object (the `key`) in another object. It was about restricting the amount of information available to the function:
```
deletePersonalAPIKeySuccess({ label })
```
would also be fine, but I'll check out the docs.

ELanning commented on frontend/src/types.ts:
```diff
@@ -12,6 +12,16 @@ export interface UserType {
+    label: string
+    value?: string
+    created_at: string
+    last_used_at: string
```
Is mixing snake case and camel/pascal case common throughout the project? That seems unfortunate.

I realize the backend is in Python which wants snake case. This can typically be solved by adding a thin middleware abstraction which converts pascal/camel case into snake case. There are other solutions as well.

Not a huge deal, but seems like a refactor worth doing eventually.

ELanning commented on posthog/api/test/test_team.py:
```diff
@@ -10,7 +10,7 @@
 
 class TestTeamUser(BaseTest):
-    TESTS_API = True
+    z = True
```
?

Twixes replied:
Misclick! Thanks.

ELanning reviewed (commented):
Didn't review much of the backend. Nice front-end work though.

Twixes commented on frontend/src/scenes/setup/PersonalAPIKeys.tsx:
```diff
@@ -0,0 +1,149 @@
+            }}
+            onCancel={() => {
+                setVisible(false)
+                inputRef.current.state.value = undefined
```
To reset input text.

Twixes committed:
Fix some code

Twixes committed:
Install git before running pip install in Dockerfiles

Twixes committed:
Improve personal API keys UI

Twixes committed:
Satisfy mypy

Twixes committed:
Reword key label placeholder

Twixes committed:
Add personal API key support to /api/user/*

Unfortunately these endpoints are still limited by CSRF protections at the moment, so not accessible outside PostHog itself.

Twixes committed:
Improve PersonalAPIKeyAuthentication and add CsrfOrKeyViewMiddleware

Twixes committed:
Merge branch 'master' into 906-personal-api-keys-and-zapier

Twixes committed:
Run collectstatic before test

Twixes committed:
Don't install dev dependencies in CI

Twixes committed:
Update dependency installation order in CI

Twixes committed:
Fix bug and describe PersonalAPIKeyAuthentication

Twixes committed:
Fix CI issues

Twixes committed:
Fix typing issues

Twixes committed:
Fix more typing issues

Twixes committed:
Merge branch 'master' into 906-personal-api-keys-and-zapier

Twixes committed:
Merge branch 'master' into 906-personal-api-keys-and-zapier

Twixes committed:
Merge branch 'master' into 906-personal-api-keys-and-zapier

Twixes committed:
Merge branch 'master' into 906-personal-api-keys-and-zapier

Twixes committed:
Use /api/personal_api_keys to list keys

Twixes committed:
Move REST hooks (and therefore Zapier) to ee/

Twixes committed:
Refactor personal API logic with kea-loaders

Twixes committed:
Add "More about API authentication in PostHog docs."

Twixes committed:
Update PersonalAPIKeys.tsx

Twixes committed:
Use TestMixin

Twixes committed:
Fix "Authentication" that should've been "Authorization"

Twixes committed:
Add option to skip self.client.force_login in API tests

Twixes committed:
Include team_id and user_id in personal API key serialization

Twixes committed:
Update test_hooks.py

Twixes committed:
Add personal API key tests

Twixes committed:
Remove leftover

Twixes committed:
Make ee.settings override posthog.settings

Twixes committed:
Merge branch 'master' into 906-personal-api-keys-and-zapier

Twixes commented:
I'm getting `No module named 'posthog.models.user.License'; 'posthog.models.user' is not a package` in license tests and not sure why, could you take a look @timgl?

timgl committed:
Don't directly import from models

timgl commented:
I made [this change](https://github.com/PostHog/posthog/pull/1281/commits/91b634f5f6e795dfdbf92c486db6f2372e1bf73b) which should fix this, although I'm not 100% sure why

Twixes committed:
Remove unused imports

Twixes committed:
Fix mypy issues

Twixes committed:
Fix HOOK_DELIVERER

Twixes committed:
Use decorator for /api/user PAK auth

Twixes committed:
Don't fire REST hook if user doesn't have "zapier" feature

Twixes committed:
Import Optional

Twixes committed:
Reword to "premium Zapier"

Twixes committed:
Make mypy happy

Twixes commented:
Ah, classic circular imports.
So this should be good now! REST hooks (used by Zapier) are only fired if the user is on a plan containing feature "zapier". Just a couple of things:
1. Shouldn't plans rather be checked on a per-team rather than per-user basis?
2. Not sure how to best convey to users Zapier-side that what they set up won't work on our free plan. Looked at how some other companies have handled it and it should be OK without any additional feature availability checking.

Twixes committed:
Fix test_delete_personal_api_key

Twixes committed:
Fix misclick

Twixes committed:
Fix and test /capture with personal API key

Twixes committed:
Make mypy happy

timgl commented on production.Dockerfile:
```diff
@@ -31,6 +27,18 @@ RUN apt-get update && apt-get install -y --no-install-recommends curl \
     && rm -rf frontend/dist/*.map
 
+# install dependencies but ignore any we don't need for dev environment
+RUN pip install $(grep -ivE "psycopg2" requirements.txt) --no-cache-dir --compile\
```
Why switch around the order of this?

Twixes replied:
Git needs to be installed before pip is ran now, as one dependency in `requirements.txt` is a `git+foo` one ([django-rest-hooks](https://github.com/zapier/django-rest-hooks) – unfortunately the version on PyPI is outdated).

timgl commented on ee/settings.py:
```diff
@@ -0,0 +1,13 @@
+
+# Zapier
+HOOK_EVENTS = {
+    # "event_name": "App.Model.Action" (created/updated/deleted)
```
Don't need this presumably?

Twixes replied:
Do you mean just `# "event_name": ...`?

Twixes committed:
Remove extra_properties_json

Twixes committed:
Merge branch 'master' into 906-personal-api-keys-and-zapier

Twixes committed:
Resolve migrations

Twixes commented:
PostHog/posthog-zapier now supports an API Host field when setting up authentication, allowing our official app to work with self-hosted solutions (though these will still be able to set up their completely own private apps as well). Event properties in Capture Event are handled there better as well.

Twixes committed:
Remove apt-utils

Twixes committed:
Optimize and test PAK user.is_active filtering

Twixes committed:
Replace DEBUG true with 1

EDsCODE reviewed (changes requested):
First, minor code comments below mainly on UI. Will do a full QA now

Error response with deleting key:
<img width="1223" alt="Screen Shot 2020-08-20 at 10 04 47 AM" src="https://user-images.githubusercontent.com/13127476/90781498-a48aff80-e2cc-11ea-8eca-47adaf52d610.png">

Twixes commented on frontend/src/scenes/setup/Setup.js:
```diff
@@ -26,21 +27,27 @@ function _Setup() {
+            variable instead of hard-coding.
+            <Divider />
+            <h2 id="personal-api-keys">Personal API Keys</h2>
+            <PersonalAPIKeys />
```
It is all kind of bundled together. I wouldn't want that to reorganize that here, but can think about that later.

Twixes commented on frontend/src/lib/components/PersonalAPIKeys/PersonalAPIKeys.tsx:
```diff
@@ -0,0 +1,163 @@
+            <p style={{ marginBottom: 0 }}>
+                Key value <b>will only ever be shown once</b>, immediately after creation.
+                <br />
+                Copy it to your destination right away.
```
Yep, good point. I think auto-copying may be a bit too intrusive, but I'll add a copy to clipboard button.

Twixes committed:
Remove unused instance_id

Twixes committed:
Improve typing

Twixes committed:
Fix deletion toast

Twixes committed:
Merge branch 'master' into 906-personal-api-keys-and-zapier

Twixes committed:
Refactor CopyToClipboard and use it in PAKs

Twixes committed:
Use toast.success

Twixes commented:
@EDsCODE Thanks, that should all be resolved now.

Twixes committed:
Merge branch 'master' into 906-personal-api-keys-and-zapier

Twixes committed:
Update migrations

Twixes committed:
Fix migration

Twixes committed:
Fix migrations

timgl commented:
Argh, couple more merge conflicts but happy to review after

Twixes committed:
Merge branch 'master' into 906-personal-api-keys-and-zapier

Twixes committed:
Complete merge

Twixes commented:
Resolved.

timgl reviewed (approved):
This looks great, well done! I'll merge this tomorrow morning and deploy
